packages/core/test-helpers/src/websocket.ts

```diff
@@ -40,41 +44,78 @@ export const MockWsServer = Server
-    let currentExchange = 0
+    const buildPayload = (msg: unknown) => (typeof msg === 'string' ? msg : JSON.stringify(msg))
+
+    // Parse requests beforehand to avoid stringifying all exchanges on every incoming ws message
+    const parsedFlow: ParsedWsMessageExchange[] = flow.map((exchange) => ({
+      ...exchange,
+      expected: buildPayload(exchange.request),
+    }))
+
     server.on('connection', async (connection) => {
-      connection.on('message', (received) => {
-        const { request, response } = flow[currentExchange]
-        let expected
-        if (typeof request === 'string') {
-          expected = request
+      // Handler to send the response (or responses) back to the client
+      const sendResponse = (response: unknown, nested = false) => {
+        if (!nested && Array.isArray(response) && response.length > 0) {
+          response.forEach((r) => sendResponse(r, true))
         } else {
-          expected = JSON.stringify(request)
+          connection.send(buildPayload(response))
         }
+      }
 
-        if (received === expected) {
-          if (Array.isArray(response) && response.length > 0) {
-            response.forEach((r) => {
-              if (typeof r === 'string') {
-                return connection.send(r)
-              }
-              return connection.send(JSON.stringify(r))
-            })
-          } else if (typeof response === 'string') {
-            connection.send(response)
-          } else {
-            connection.send(JSON.stringify(response))
-          }
+      connection.on('message', (received) => {
+        let exchange
+
+        if (parsedFlow.length === 0 && options.errorOnUnexpectedMessage) {
+          throw Error(`Unexpected WS message, received: '${received}'`)
+        }
 
-          currentExchange++
-          if (currentExchange === flow.length) {
-            resolve(true)
+        if (options.enforceSequence) {
+          // If the sequence is enforced, get first item from the flow of exchanges
+          exchange = parsedFlow[0] as ParsedWsMessageExchange
+
+          if (exchange.expected !== received) {
+            if (options.errorOnUnexpectedMessage) {
+              throw Error(
+                `The WS message received does not match the expected one.
+                 Expected: '${exchange.expected}'
+                 Received: '${received}'`,
+              )
+            } else {
+              return
+            }
           }
+
+          // Message received matches expected request, remove exchange from the list
+          parsedFlow.shift()
         } else {
-          throw Error(
-            `The WS message received does not match the expected one. \nExpected: ${expected}\nReceived${received}`,
-          )
+          // If the sequence is not enforced, try to find the received msg within the list of expected exchanges
+          const i = parsedFlow.findIndex((exchange) => exchange.expected === received)
+
+          if (i === -1) {
```

Comment:
Personally I like using `.some()` better when the index itself doesn't matter:
```suggestion
          const messageExpected = parsedFlow.some((exchange) => exchange.expected === received)

          if (!messageExpected) {
```

Reply:
it uses index few lines below to remove found element. If we get rid of index we would use something like filter which is less performant in this case

Reply:
Oh shoot, you're right. Ignore this comment...